src/librustc/traits/error_reporting.rs

```diff
@@ -466,6 +466,20 @@ pub fn report_selection_error<'a, 'tcx>(infcx: &InferCtxt<'a, 'tcx>,
                         err.emit();
                     }
 
+                    ty::Predicate::ClosureKind(closure_def_id, kind) => {
+                        let found_kind = infcx.closure_kind(closure_def_id).unwrap();
+                        let closure_span = infcx.tcx.map.span_if_local(closure_def_id).unwrap();
+                        let mut err = struct_span_err!(
+                            infcx.tcx.sess, closure_span, E0524,
+                            "the closure implements `{}` but not `{}`",
```

Comment:
Nit: I think you should say something like "expected a closure that implements the `{}` trait, but this closure only implement `{}`". This fits our usual "expected, found" order better, and helps clarify that `FnMut` (etc) is a trait.